[WIP] First version of G2EGM.

pkofod opened the pull request:
Comments, issues, and ideas will follow, but feel free to review.

I'm having some issues with *acon* still, the segment where all liquid resources are consumed or spent (no liquid savings). I've asked Jeppe for a meeting to discuss what might be going on here.

I should say that some things are on my list, but have been put lower on the priority list due to the *acon* problems. Such things are: don't hardcode `C`, `D` etc as parameters, but pass pass lists or higher dimensional arrays as appropriate (see `cleanSegment` for an example that doesn't require four specific arrays to be passed), don't use classes for "stablevalue"-things, use functions (I believe such a thing exists in HARK already? couldn't figure out if it was example specific) instead, and more.

The retirement is a standard EGM exercise. Should there be a separate HARK implementation for this?

llorracc commented:
What kinds of problems? 

Probably related to one or both of the kink points?  And maybe how those interact with the discrete choice?

pkofod commented:
Consider these `imshow` pllots of the discrete policies (yellow is work)
![good](https://user-images.githubusercontent.com/8431156/47049629-2d4f2680-d19e-11e8-933b-66b5651d394f.png)

These will be the solutions if you fully ignore the `acon` segment. Then consider uncommenting the line that is currently commented out to ignore `acon` and you get

![bad](https://user-images.githubusercontent.com/8431156/47049663-48219b00-d19e-11e8-8c6f-9deda574a9e6.png)


So either I'm not filtering out the infeasible policies and or endogenous states correctly (c, d, m and n cannot endogenously become < 0.0 and still be feasible solutions to the FOCs), calculating the FOC candidates correctly, or something else. I'm hoping Jeppe and Thomas have seen similar results before and know what the solution is. I believe they do some special handling of infeasible points that I don't, and that might be the answer, but I don't know.

pkofod commented:
Regarding honoring the NARK, should risk free return factors then follow the `Rfree` notation somehow? There are two of them, since both investment in `A` and `B` is risk free.

pkofod commented:
I've added some additional commits. Few of them were directly related to the `acon` thing, but I just had a look, and it seems clear that this comes from the density of the endogenous grid in (m,n) space.

**Coarse grid**
A too coarse (as in this PR) grid over post decision cash on hand gives policies as above and (m,n) grids as shown below (first one shows all endogenous grid points, second one is zoomed in).

![coarsebiggrid](https://user-images.githubusercontent.com/8431156/47115242-40293000-d25e-11e8-9b41-259a4d43b479.png)
![coarsesmallgrid](https://user-images.githubusercontent.com/8431156/47115245-41f2f380-d25e-11e8-9392-25ef1eabdbe8.png)
![coarsepolicy](https://user-images.githubusercontent.com/8431156/47115249-44ede400-d25e-11e8-8c31-eee47bd4395f.png)

The artifacting is there and the grid is seen to be very coarse in the lower part of (m,n) space).

** Fine grid **
![finebiggrid](https://user-images.githubusercontent.com/8431156/47115294-6fd83800-d25e-11e8-84a2-13e3dd6304ee.png)
![finesmallgrid](https://user-images.githubusercontent.com/8431156/47115296-71a1fb80-d25e-11e8-86ec-44f809af8a6b.png)
![finepolicy](https://user-images.githubusercontent.com/8431156/47115300-736bbf00-d25e-11e8-9358-f4036f18fa6d.png)

the problem seems to go away.

llorracc commented:
> Regarding honoring the NARK, should risk free return factors then follow the `Rfree` notation somehow? There are two of them, since both investment in `A` and `B` is risk free.

Yes, please use Rfree.  And, generically, there is only ever a single risk free interest rate because if there were two all savers would choose the higher and all borrowers would choose the lower so saving would not match borrowing.

llorracc commented:
You're describing these results as 'artifacts' which is ambiguous as to whether they might actually be correct solutions to the problem as specified.  The problem itself has discontinuities in behavior, and those can interact with discrete probability distributions in funny ways.  Have you tried solving with the coarse grid that has the artifacts back to some period 0, then constructing a superfine grid for the prior period and solving that?  If so, and you get the same results as with the coarser grid, the result is probably reflecting a correct result to a problem that has some ugly little corners.

pkofod commented:
> Yes, please use Rfree. And, generically, there is only ever a single risk free interest rate because if there were two all savers would choose the higher and all borrowers would choose the lower so saving would not match borrowing.

Okay, then maybe my terminology is wrong.

There are two assets: `A` and `B`. If you have `a` units of `A` at the end of period `t`, this will grow to `a*R_a` in period `t+1`. Similarly, if you have `b` units of `B` at the end of period `t`, this will grow to `b*R_b` in period `t+1`. This is why I'm calling them risk free. You invest one unit and you get `R_x` units back 100% guaranteed - no chance it'll be anything else.

However, `A` is fully liquid and `B` is illiquid in the sense that you can only consume it (or accumulate it in `A` if you wanted to do that) if you retire. `R_a < R_b`, so people want to invest (depending on the current state of course) in both.

Is `R_b` not the return factor of a risk free (although illiquid) asset in this description, or is there a better term than "risk free"?

pkofod commented:
> You're describing these results as 'artifacts' which is ambiguous as to whether they might actually be correct solutions to the problem as specified. The problem itself has discontinuities in behavior, and those can interact with discrete probability distributions in funny ways. Have you tried solving with the coarse grid that has the artifacts back to some period 0, then constructing a superfine grid for the prior period and solving that? If so, and you get the same results as with the coarser grid, the result is probably reflecting a correct result to a problem that has some ugly little corners.

I'm going to push some more commits in the next half hour, but I'm quite certain that it's artifacting in "a bad way" :) I'm certain because I can make them go away simply by increasing precision in the sense that I add more points and interpolate on a finer grid. This, to me, shows that it's not a discontinuity in the solution (they're in there too!). It's basically related to some of the things discussed around pages 94-96 in the paper.

mnwhite commented:
This PR can be removed, given the other DCEGM PR's are much more complete, yes?

pkofod commented:
Did you mean to write this comment on the DCEGM PR? #206

mnwhite commented:
G2EGM got mashed into DCEGM in my brain.  Ignore.

pkofod commented:
I want this to continue living, and come back and revisit some of the things in the very near future when DCEGM is settled.

llorracc commented:
@pkofod, could you update the status of this, now that DCEGM is incorporated in core HARK?

pkofod commented:
DCEGM is in HARK as a tool that you can use to build a model of mixed discrete and continuous choices... If the EGM step involves 1 post-decision state. It seems to work quite well. G2EGM can also solve the same kinds of models. The "1 post-decision state" case is still solved in a different kind of way than DCEGM though, so G2EGM is not a generalization of DCEGM.

For models that involve 2 or more post-decision states, we cannot use DCEGM anymore (though we can actually solve a series of problems over one state given values of the other state). G2EGM can in principle be used in as high dimensions as you'd like, provided you want to write (or can automate the code generation of) the interpolation code for the number of dimensions you have.

This code is current in the state that the DCEGM code was in before I rewrote it, and split the model out into a REMARK and the "how does it work" out into a DemARK.

I think it's still worth doing, but I think the things that are explicitly linked to the TFI goals are more imporant - unless we end up having to solve a higher dimensional problem.

I could quite easily extract the G2EGM out just like I did with DCEGM, but the DemARK and REMARK parts would take a bit of time to get right.

llorracc commented:
@Mv77, 

In our meeting today we were looking at this old PR which never got finished.  Could you look at it before our call next Monday and we can discuss whether you might be able to polish it off.

Mv77 commented:
> 
> 
> @Mv77,
> 
> In our meeting today we were looking at this old PR which never got finished. Could you look at it before our call next Monday and we can discuss whether you might be able to polish it off.

Sure, I'll take a look and be up to speed on it by our next call.

Mv77 commented:
I have been revising/using the code and wanted to write an update.

I thought a good way to move forward would be to start the sketch of a REMARK of the G2EGM paper, trying to replicate it. In particular they have these plots where they show which segment is optimal on different parts of the state space. I produced a tool to get similar plots. Here is an example:

The original:
![image](https://user-images.githubusercontent.com/27739595/88339763-9f0bbb00-cd08-11ea-85c1-9eadd21b76f6.png)
HARK's:
![image](https://user-images.githubusercontent.com/27739595/88339794-aa5ee680-cd08-11ea-8fa4-3361a0321cd8.png)

There is some resemblance and the segments' optimality regions are in the same general directions. However there puzzles that I am currently investigating:
- dcon is never being highlighted as the optimal segment. Its region seems to be taken over by 'con'.
- there are regions to the right where 'con' becomes optimal that shouldn't be there.

Mv77 commented:
Retirement/Working decisions look very reasonable nonetheless:
![image](https://user-images.githubusercontent.com/27739595/88340894-9a480680-cd0a-11ea-9728-d2c48a9a4524.png)
You retire when you have enough resources, or when your liquid assets (M) are low enough that you retire just to tap into your retirement account.

Mv77 commented:
There is definitely something going on with dcon. If I set the parameters to make asset b bad (Rb = 0.5, chi very small), then the agent should never want to put resources into it, and so the d constraint should always bind. Thus, in principle, 'ucon' should be dominated by 'dcon'.

However, I still get solutions where dcon is never optimal.

Mv77 commented:
Update:
I've found what I think are two bugs.

- The fixed C's in acon need to be transposed: they are generated row-wise using limits that make sense column-wise. The current version therefore generates many C-b combinations that imply negative d's and are thus thrown out. This is fixed in https://github.com/Mv77/HARK/commit/62c2e20a6dffafa9cb158cd47965e3dc9355b812. I think it might help with some of the previously documented issues with acon.

- There indeed seems to be a major issue with dcon: in my tests, the denominator of the barycentric weights for this segment is always 0! This results in its values never being passed onto the common grid. I'm in the process of figuring out why this happens with this particular segment.

llorracc commented:
Hmmm, a denominator of zero does sound like a bit of a problem!

Sounds like you are pursuing the strategy of using the tool to solve a
simple problem for which you already know the answer?  (The "rettirement"
problem?).  That's always a good debugging strategy!


On Fri, Jul 24, 2020 at 6:09 PM Mateo Velásquez-Giraldo <
notifications@github.com> wrote:

> Update:
> I've found what I think are two bugs.
>
>    -
>
>    The fixed C's in acon need to be transposed: they are generated
>    row-wise using limits that make sense column-wise. The current version
>    therefore generates many C-b combinations that imply negative d's and are
>    thus thrown out. This is fixed in Mv77@62c2e20
>    <https://github.com/Mv77/HARK/commit/62c2e20a6dffafa9cb158cd47965e3dc9355b812>.
>    I think it might help with some of the previously documented issues with
>    acon.
>    -
>
>    There indeed seems to be a major issue with dcon: in my tests, the
>    denominator of the barycentric weights for this segment is always 0! This
>    results in its values never being passed onto the common grid. I'm in the
>    process of figuring out why this happens with this particular segment.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/econ-ark/HARK/pull/200#issuecomment-663753887>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKCK74WLJLKFFDW54UZDDDR5IBDLANCNFSM4F4JRNRA>
> .
>


-- 
- Chris Carroll

Mv77 commented:
Yes! I am doing that and also trying to match the solution of the model in the original paper.

The zero denominator issue was that the interpolator was being passed (M,M) instead of (M,N). This resulted in the intended 2-d mesh to just be a straight 45° line. No triangulation can be done across a line and this manifested itself in 0 denominators. That issue is fixed now.

sbenthall commented:
This PR is superseded by #782 . @Mv77 is going to bring this home.
